blockchain/utxocache.go

```diff
@@ -0,0 +1,793 @@
+	// loaded.
+	tfModified
+
+	// tfFresh indicates that the entry is fresh.  This means that the parent
+	// view never saw this entry.  Note that tfFresh is a performance
+	// optimization with which we can erase entries that are fully spent if we
+	// know we do not need to commit them.  It is always safe to not mark
+	// tfFresh if that condition is not guaranteed.
+	tfFresh
+)
+
+// UtxoEntry houses details about an individual transaction output in a utxo
+// view such as whether or not it was contained in a coinbase tx, the height of
+// the block that contains the tx, whether or not it is spent, its public key
+// script, and how much it pays.
+type UtxoEntry struct {
+	// NOTE: Additions, deletions, or modifications to the order of the
+	// definitions in this struct should not be changed without considering
+	// how it affects alignment on 64-bit platforms.  The current order is
+	// specifically crafted to result in minimal padding.  There will be a
+	// lot of these in memory, so a few extra bytes of padding adds up.
+	// Any changes here should also be reflected in the memoryUsage() function.
+
+	amount      int64
+	pkScript    []byte // The public key script for the output.
+	blockHeight int32  // Height of block containing tx.
+
+	// packedFlags contains additional info about output such as whether it
+	// is a coinbase, whether it is spent, and whether it has been modified
+	// since it was loaded.  This approach is used in order to reduce memory
+	// usage since there will be a lot of these in memory.
+	packedFlags txoFlags
+}
+
+// IsCoinBase returns whether or not the output was contained in a coinbase
+// transaction.
+func (entry *UtxoEntry) IsCoinBase() bool {
+	return entry.packedFlags&tfCoinBase == tfCoinBase
+}
+
+// IsSpent returns whether or not the output has been spent based upon the
+// current state of the unspent transaction output view it was obtained from.
+func (entry *UtxoEntry) IsSpent() bool {
+	return entry.packedFlags&tfSpent == tfSpent
+}
+
+// isModified returns whether or not the output has been modified since it was
+// loaded.
+func (entry *UtxoEntry) isModified() bool {
+	return entry.packedFlags&tfModified == tfModified
+}
+
+// isFresh returns whether or not it's certain the output has never previously
+// been stored in the database.
+func (entry *UtxoEntry) isFresh() bool {
+	return entry.packedFlags&tfFresh == tfFresh
+}
+
+// BlockHeight returns the height of the block containing the output.
+func (entry *UtxoEntry) BlockHeight() int32 {
+	return entry.blockHeight
+}
+
+// Amount returns the amount of the output.
+func (entry *UtxoEntry) Amount() int64 {
+	return entry.amount
+}
+
+// PkScript returns the public key script for the output.
+func (entry *UtxoEntry) PkScript() []byte {
+	return entry.pkScript
+}
+
+// memoryUsage returns the memory usage in bytes of the UTXO entry.
+// It returns 0 for the nil element.
+func (entry *UtxoEntry) memoryUsage() uint64 {
+	if entry == nil {
+		return 0
```

Comment:
I realize nil's aren't free. At first I thought it wouldn't work easy with adding and removing them, but I think right now it might. Let me make sure. Thanks for starting to look at this, though!

Reply:
I think it might make sense to instead have `totalMemoryUsage` add 8 bytes per entry.

Reply:
I went for that approach. Let me know if you think it makes more sense to add the 8 here instead of in the other method.